### ISSUE TITLE
Deallocations within zstd violate GlobalAlloc's safety contract

### ISSUE THREAD

**Issue by SFBdragon** [gyscos/zstd-rs#248]
Hi!

(All testing was done on WASM.)

Here's the [documentation](https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html#tymethod.dealloc) for GlobalAlloc's `dealloc`. In particular:

> `layout` must be the same layout that was used to allocate that block of memory.

The C code in `zstd-rs` breaks this, try running the following code in a WASM project (I think you could adapt this and see a similar result on desktop platforms, but I haven't confirmed this).

lib.rs
```rust
use wasm_bindgen::prelude::*;

struct BadAllocator;

unsafe impl std::alloc::GlobalAlloc for BadAllocator {
    unsafe fn alloc(&self, layout: std::alloc::Layout) -> *mut u8 {
        let ptr = (std::arch::wasm32::memory_grow::<0>(layout.size() / 64 / 1024 + 1) * 64 * 1024) as *mut u8;

        log("alloc: ptr, size");
        log_u32(ptr as u32);
        log_u32(layout.size() as u32);

        ptr
    }

    unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
        log("dealloc: ptr, size");
        log_u32(ptr as u32);
        log_u32(layout.size() as u32);
    }
}

#[global_allocator]
static BADALLOC: BadAllocator = BadAllocator;


#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = console)]
    fn log(s: &str);

    #[wasm_bindgen(js_namespace = console, js_name = log)]
    fn log_u32(a: u32);
}

#[wasm_bindgen]
pub fn compress(data: &[u8], level: i32) -> Vec<u8> {
    zstd::bulk::compress(data, level).unwrap()
}
```

index.js
```js
import * as wasm from "wasm-alloc-bench";

var array = new Array(10000);

for (let i = 0; i < array.length; i++) {
    array[i] = i % 256;
}

wasm.compress(array, 0);
```

Which produces the output:

![console output](https://github.com/gyscos/zstd-rs/assets/46199751/a3ab0ac5-30fe-42a9-bfd6-c52af5056bba)

Notice that there are two deallocations of size 1, which were allocated with an entirely different size (`layout` contains the size).

So why does the code normally work?
- The default WASM allocator, `dlmalloc` is a direct port/wrapper of the C code, which uses the C malloc/free API, which does not take a size for the deallocation. This allocator stores that information in the arena, so giving it a wrong size via the GlobalAlloc interface doesn't break things.
- `lol_alloc` is another popular WASM allocator. This one doesn't run into errors in this case, but it ends up only freeing a tiny amount of space each deallocation. This means that repeated calls to `zstd-rs` functions using `lol_alloc` will result in a lot of leaked memory which is impossible to reclaim.

My crate, [`talc`](https://github.com/SFBdragon/talc) relies on this invariant, and throws an error when users try to use this crate with mine. See the issue [here](https://github.com/SFBdragon/talc/issues/17).

**Comment by gyscos** [comment]
Thanks for the report!

I suspect it's related to this line:
https://github.com/gyscos/zstd-rs/blob/d84bac590aa12f40e0eb1722593f9d10690eb541/zstd-safe/zstd-sys/src/wasm_shim.rs#L25

Which might be fine with the default allocator, but that might indeed not be true for others.

I'm not super familiar with this part of the code, so I'm open to suggestions.

**Comment by SFBdragon** [comment]
No problem!

Yep, that looks right. 

I'm lean on suggestions, though. Getting the size information for the deallocation means patching the C code to use a different function or storing a map between allocated pointers and their sizes on the Rust side - ick.

**Comment by gyscos** [comment]
> storing a map between allocated pointers and their sizes on the Rust side

That's what the default C allocator already does I assume?

**Comment by SFBdragon** [comment]
Usually not. It generally makes more sense to keep the size data next to the allocation, in the heap, or something to that effect. It's simpler, faster, and usually more memory efficient. (Not as good as just having the caller give you the size directly!) Default C allocators vary though, I'm not familiar with the nuances.

My main source of hesitation is just that it'd be global state, which isn't ideal, but it seems like the most viable option thus far.

**Comment by SFBdragon** [comment]
Oh, perhaps we can change those FFI bindings to handle that. Allocate slightly more memory than demanded, write the size next to the rest of the data, and read it back when freeing accordingly.

(I really need some shut-eye for now. I should be more coherent tomorrow, heh)

**Comment by gyscos** [comment]
Hmmm yeah we want something so that:
* The actual allocation is big enough to include the original request + the layout we want to store (size and alignment).
* We'll return a pointer to the part of the allocation that stores the original requested layout. We need to be able to go to the layout storage from there, _without knowing the layout_. This pretty much means we can only store the layout _right before_ the requested allocation. `[size] [alignment] [original requested layout]`.
* Because of alignment, this might not be the start of the actual allocation. But since we have the size and alignment of the original layout, we can re-compute the "padded layout", and determine the start of the actual allocation.
* We can then free that.

**Comment by SFBdragon** [comment]
Pretty much. There's code that does similar stuff in my allocator, but this solution is more complex than is needed, especially as malloc/calloc cannot request arbitrary alignment. I've implemented the fix by adding a few lines to the wasm_shim file and I've tested it with the example code I provided and by allocator, and everything seems to be in order.

Before I open a PR though, I couldn't build the crate without switching all the references in zstd_sys from std to core and importing the alloc crate, as the library contains the `#[no_std]` directive. However, I assume the code has been working for you, so would you rather I exclude the changes to std/core references or leave them out?

**Comment by gyscos** [comment]
> Before I open a PR though, I couldn't build the crate without switching all the references in zstd_sys from std to core and importing the alloc crate, as the library contains the #[no_std] directive. However, I assume the code has been working for you, so would you rather I exclude the changes to std/core references or keep them in the commit?

Whoops there used to be a `std` features that re-imported std, but [that has now been removed](https://github.com/gyscos/zstd-rs/commit/025bd7dc0f1d32b7612509ad9e09b45fd1a5e180#diff-3310e036a6697701b314a4dfb331e76335e5edc0c94def162808d64b5b2fa707L9-L10) (since everything we needed is now in libcore), so at the moment it seems indeed that this cannot build.

I'll add back the import, so at least it can be built with that `std` feature.

**Comment by SFBdragon** [comment]
Hmm, I don't know the ins and outs of the build system, but I've got a patch file with my changes that include switching everything to using core if you'd rather not reintroduce the feature and keep zstd_sys as general as possible. Let me know.

**Comment by gyscos** [comment]
Using libcore for `c_int` and `c_void` is easy, but for `std::alloc::alloc` (and the zeroed version) we'll likely need the `alloc` crate.

I don't know of any negative impact of using this instead of `std`, so let's try that?

**Comment by SFBdragon** [comment]
Implemented the change, but [loadshedding](https://en.wikipedia.org/wiki/South_African_energy_crisis) hit, oops.

In principle, `std` just reimports the stuff from for and `alloc`. There used to be some kerfuffle about how cargo handled prebuilt binaries and std in particular, but I think that's been resolved for a while. Regardless, looks like you've cleaned up the loose ends (not sure what that module import was, my bad). Thanks!

**Comment by gyscos** [comment]
Published zstd-0.13.0 with this fix.

**Comment by SFBdragon** [comment]
v0.13 is working on my end :)
